pkg/cluster/internal/create/actions/join/join.go

```diff
@@ -0,0 +1,267 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package config implements the kubeadm config action
+package config
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"sigs.k8s.io/kind/pkg/cluster/constants"
+	"sigs.k8s.io/kind/pkg/cluster/internal/create/actions"
+	"sigs.k8s.io/kind/pkg/cluster/internal/kubeadm"
+	"sigs.k8s.io/kind/pkg/cluster/nodes"
+	"sigs.k8s.io/kind/pkg/exec"
+	"sigs.k8s.io/kind/pkg/fs"
+)
+
+// Action implements action for creating the kubeadm config
+// and deployng it on the bootrap control-plane node.
+type Action struct{}
+
+// NewAction returns a new action for creating the kubadm config
+func NewAction() actions.Action {
+	return &Action{}
+}
+
+// Execute runs the action
+func (a *Action) Execute(ctx *actions.ActionContext) error {
+	allNodes, err := ctx.Nodes()
+
+	// join secondary control plane nodes if any
+	secondaryControlPlanes, err := nodes.SecondaryControlPlaneNodes(allNodes)
+	if len(secondaryControlPlanes) > 0 {
+		if err := joinSecondaryControlPlanes(
+			ctx, allNodes, secondaryControlPlanes,
+		); err != nil {
+			return err
+		}
+	}
+
+	// then join worker nodes if any
+	workers, err := nodes.SelectNodesByRole(allNodes, constants.WorkerNodeRoleValue)
+	if err != nil {
+		return err
+	}
+	if len(workers) > 0 {
+		if err := joinWorkers(ctx, allNodes, workers); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func joinSecondaryControlPlanes(
+	ctx *actions.ActionContext,
+	allNodes []nodes.Node,
+	secondaryControlPlanes []nodes.Node,
+) error {
+	ctx.Status.Start("Joining control-plane nodes to Kubernetes ☸")
+	defer ctx.Status.End(false)
+
+	// TODO(bentheelder): this should be concurrent
+	for _, node := range secondaryControlPlanes {
+		if err := runKubeadmJoinControlPlane(ctx, allNodes, &node); err != nil {
+			return err
+		}
+	}
+
+	ctx.Status.End(true)
+	return nil
+}
+
+func joinWorkers(
+	ctx *actions.ActionContext,
+	allNodes []nodes.Node,
+	workers []nodes.Node,
+) error {
+	ctx.Status.Start("Joining worker nodes to Kubernetes ☸")
+	defer ctx.Status.End(false)
+
+	// TODO(bentheelder): this should be concurrent
```

Comment:
unlike joining CPs, this shouldn't be blocked on CM locks, but i haven't tested.

Reply:
I have a WIP patch to make these concurrent.